sled-agent/src/instance_manager.rs

```diff
@@ -775,6 +800,53 @@ impl InstanceManagerRunner {
         tx.send(Ok(state)).map_err(|_| Error::FailedSendClientClosed)?;
         Ok(())
     }
+
+    async fn only_use_disks(
+        &mut self,
+        tx: oneshot::Sender<Result<(), Error>>,
+        disks: AllDisks,
+    ) -> Result<(), Error> {
+        // Consider the generation number on the incoming request to avoid
+        // applying old requests.
+        let requested_generation = *disks.generation();
+        if let Some(last_gen) = self.storage_generation {
+            if last_gen >= requested_generation {
+                // This request looks old, ignore it.
+                info!(self.log, "only_use_disks: Ignoring request";
+                    "last_gen" => ?last_gen, "requested_gen" => ?requested_generation);
+
+                return Ok(());
+            }
+        }
+        self.storage_generation = Some(requested_generation);
+        info!(self.log, "only_use_disks: Processing new request";
+            "gen" => ?requested_generation);
+
+        let u2_set: HashSet<_> = disks.all_u2_zpools().into_iter().collect();
+
+        let mut to_remove = vec![];
+        for (id, (_, instance)) in self.instances.iter() {
+            // If we can read the filesystem pool, consider it. Otherwise, move
+            // on, to prevent blocking the cleanup of other instances.
+            let Ok(Some(filesystem_pool)) =
+                instance.get_filesystem_zpool().await
+            else {
+                info!(self.log, "only_use_disks: Cannot read filesystem pool"; "id" => ?id);
+                continue;
+            };
+            if !u2_set.contains(&filesystem_pool) {
+                to_remove.push(*id);
+            }
+        }
+
+        for id in to_remove {
+            info!(self.log, "only_use_disks: Removing instance"; "id" => ?id);
```

Comment:
similarly, i have a slight preference for keeping the field name consistent even though i'll admit the message makes it clear what it's the ID of here:
```suggestion
            info!(self.log, "only_use_disks: Removing instance"; "instance_id" => ?id);
```

Reply:
Updated